crates/store/src/state.rs

```diff
@@ -777,64 +684,135 @@ impl State {
+
+        // SAFETY: This should not error because:
+        // - we're passing exactly the block headers that we've added to the partial MMR,
+        // - so none of the block header's block numbers should exceed the chain length of the
+        //   partial MMR,
+        // - and we've added blocks to a BTreeSet, so there can be no duplicates.
+        let chain_mmr = ChainMmr::new(partial_mmr, headers)
+            .expect("partial mmr and block headers should be consistent");
+
+        Ok(BlockInputs::new(
+            latest_block_header,
+            chain_mmr,
+            account_witnesses,
+            nullifier_witnesses,
+            unauthenticated_note_proofs,
+        ))
+    }
+
+    /// Get account and nullifier witnesses for the requested account IDs and nullifier as well as
+    /// the [`PartialMmr`] for the given blocks. The MMR won't contain the latest block and its
+    /// number is removed from `blocks` and returned separately.
+    ///
+    /// This method acquires the lock to the inner state and does not access the DB so we release
+    /// the lock asap.
+    async fn get_block_inputs_witnesses(
+        &self,
+        blocks: &mut BTreeSet<BlockNumber>,
+        account_ids: Vec<AccountId>,
+        nullifiers: Vec<Nullifier>,
+    ) -> Result<
+        (
+            BlockNumber,
+            BTreeMap<AccountId, AccountWitness>,
+            BTreeMap<Nullifier, NullifierWitness>,
+            PartialMmr,
+        ),
+        GetBlockInputsError,
+    > {
+        let inner = self.inner.read().await;
+
+        let latest_block_number = inner.latest_block_num();
+
+        // If `blocks` is empty, use the latest block number which will never trigger the error.
+        let highest_block_number = blocks.last().copied().unwrap_or(latest_block_number);
+        if highest_block_number > latest_block_number {
+            return Err(GetBlockInputsError::UnknownBatchBlockReference {
+                highest_block_number,
+                latest_block_number,
             });
         }
 
-        // using current block number gets us the peaks of the chain MMR as of one block ago;
-        // this is done so that latest.chain_root matches the returned peaks
-        let chain_peaks =
-            inner.blockchain.peaks_at(latest.block_num().as_usize()).map_err(|error| {
-                GetBlockInputsError::FailedToGetMmrPeaksForForest {
-                    forest: latest.block_num().as_usize(),
-                    error,
-                }
-            })?;
-        let account_states = account_ids
+        // The latest block is not yet in the chain MMR, so we can't (and don't need to) prove its
+        // inclusion in the chain.
+        blocks.remove(&latest_block_number);
+
+        // Fetch the partial MMR at the state of the latest block with authentication paths for the
+        // provided set of blocks.
+        let partial_mmr = inner.blockchain.partial_mmr_from_blocks(blocks, latest_block_number);
```

Comment:
I had a related comment in another PR: we could probably just pass the `block` into `partial_mmr_from_blocks()` and remove the `lated_block_number` after - e.g., something like this:
```rust
let partial_mmr = inner.blockchain.partial_mmr_from_blocks(blocks);
blocks.remove(&latest_block_number);
```

Reply:
`latest_block_number` is used as the target forest of the partial MMR, so if the `blocks` do contain the latest block number and it is not removed, we will try to get a proof for block `latest_block_number` even though the MMR we're getting it from only has length `latest_block_number`, which will result in an error.